src/Symfony/Component/HttpKernel/RealHttpKernel.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Mime\Part\AbstractPart;
+use Symfony\Component\Mime\Part\DataPart;
+use Symfony\Component\Mime\Part\Multipart\FormDataPart;
+use Symfony\Component\Mime\Part\TextPart;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * An implementation of a Symfony HTTP kernel using a "real" HTTP client.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class RealHttpKernel implements HttpKernelInterface
+{
+    private $client;
+    private $logger;
+
+    public function __construct(HttpClientInterface $client = null, LoggerInterface $logger = null)
+    {
+        if (!class_exists(HttpClient::class)) {
+            throw new \LogicException(sprintf('You cannot use "%s" as the HttpClient component is not installed. Try running "composer require symfony/http-client".', __CLASS__));
+        }
+
+        $this->client = $client ?? HttpClient::create();
+        $this->logger = $logger ?? new NullLogger();
+    }
+
+    public function handle(Request $request, $type = HttpKernelInterface::MASTER_REQUEST, $catch = true)
+    {
+        $this->logger->debug(sprintf('Request: %s %s', $request->getMethod(), $request->getUri()));
+
+        $headers = $this->getHeaders($request);
+        $body = '';
+        if (null !== $part = $this->getBody($request)) {
+            $headers = array_merge($headers, $part->getPreparedHeaders()->toArray());
+            $body = $part->bodyToIterable();
+        }
+        $response = $this->client->request($request->getMethod(), $request->getUri(), [
+            'headers' => $headers,
+            'body' => $body,
+            'max_redirects' => 0,
+        ] + $request->attributes->get('http_client_options', []));
```

Comment:
This line is the only difference with #30602: an `http_client_options` attribute to allow passing options to the underlying HTTP client when needed.